scripts/util.py

```diff
@@ -153,6 +155,66 @@
 ] = lambda self, data: yaml.representer.SafeRepresenter.represent_str(self, str(data))
 
 
+class ApiEndpoint(Enum):
+    COMPUTE = "compute"
+    BQ = "bq"
+    STORAGE = "storage"
+    TPU = "tpu"
+    SECRET = "secret_manager"
+
+
+_ud = "googleapis.com"
+
+
+def universe_domain(refresh: bool = False) -> str:
+    global _ud
+    if refresh:
+        _ud = instance_metadata("attributes/universe_domain")
+    return _ud
+
+
+def endpoint_version(api: ApiEndpoint):
+    if api.value in lkp.endpoint_versions:
+        return lkp.endpoint_versions[api.value]
+    return None
+
+
+def get_credentials(credentials=None) -> service_account.Credentials:
+    try:
+        key_path = os.environ["GOOGLE_APPLICATION_CREDENTIALS"]
+    except KeyError:
+        key_path = None
+    if key_path is not None:
+        credentials = service_account.Credentials.from_service_account_file(
+            key_path, scopes=[f"https://www.{universe_domain()}/auth/cloud-platform"]
+        )
+    elif credentials is None:
+        credentials = def_creds
+
+    return credentials
+
+
+def create_client_options(api: ApiEndpoint = None) -> ClientOptions:
+    """Create client options for cloud endpoints"""
+    ep = None
+    ver = endpoint_version(api)
+    ud = universe_domain()
+    if ver:
+        ep = f"https://{api.value}.{ud}/"
+        if api == ApiEndpoint.STORAGE:
+            os.environ["_API_VERSION_OVERRIDE_ENV_VAR"] = ver
```

Comment:
Is there any doc that explains why storage is special?

Reply:
Nothing that I found immediately.  I had to dig around in the library to find that variable being set.  From first glance, it would seem that the API libraries are not consistent with how they treat the ClientOptions.  For the most part this should not affect Slurm-GCP as most of its use is with compute.

Reply:
Technically this code is unreachable with the current terraform only allowing an endpoint for compute to be defined, so I will remove it for now.